File: reference/python/mention.py
from __future__ import annotations

import re
from dataclasses import dataclass

from .config import MyConversationConfig, should_accept_group, should_require_mention, user_id_matches

MENTION_TOKEN_RE = re.compile(r"\[\[@([^:\]]+):(\d+)\]\]")


def strip_mention_tokens_for_command_detection(content: str) -> str:
    collapsed = MENTION_TOKEN_RE.sub(" ", content)
    return " ".join(collapsed.split()).strip()


def strip_leading_bot_username(content: str, username: str | None) -> str:
    text = content.strip()
    if not text or not username:
        return text
    uname = username.strip().lstrip("@")
    lowered = text.lower()
    for prefix in (f"@{uname}", uname):
        if lowered.startswith(prefix.lower()):
            return text[len(prefix) :].strip()
    return text


def looks_like_slash_command(body: str) -> bool:
    trimmed = body.strip()
    if not trimmed.startswith("/"):
        return False
    parts = trimmed.split(maxsplit=1)
    raw = parts[0][1:].lower() if parts else ""
    if not raw:
        return False
    if "@" in raw:
        raw = raw.split("@", 1)[0]
    if "/" in raw:
        return False
    return True


def detect_inbound_control_command(raw_body: str, username: str | None = None) -> bool:
    trimmed = raw_body.strip()
    if not trimmed:
        return False
    if looks_like_slash_command(trimmed):
        return True
    without_tokens = strip_mention_tokens_for_command_detection(trimmed)
    if without_tokens and without_tokens != trimmed and looks_like_slash_command(without_tokens):
        return True
    if username:
        without_user = strip_leading_bot_username(without_tokens or trimmed, username)
        if without_user and looks_like_slash_command(without_user):
            return True
    return False


def extract_command_body(raw_body: str, username: str | None = None) -> str | None:
    trimmed = raw_body.strip()
    candidates = [trimmed]
    without_tokens = strip_mention_tokens_for_command_detection(trimmed)
    if without_tokens:
        candidates.append(without_tokens)
    if username:
        candidates.append(strip_leading_bot_username(without_tokens or trimmed, username))
    for candidate in candidates:
        if candidate and looks_like_slash_command(candidate):
            return candidate
    return None


@dataclass
class MentionGateResult:
    should_skip: bool
    reason: str
    effective_was_mentioned: bool


def format_mention(display_name: str, user_id: int | str) -> str:
    return f"[[@{display_name}:{user_id}]]"


def extract_mentioned_user_ids(content: str) -> list[int]:
    seen: set[int] = set()
    ids: list[int] = []
    for match in MENTION_TOKEN_RE.finditer(content):
        user_id = int(match.group(2))
        if user_id not in seen:
            seen.add(user_id)
            ids.append(user_id)
    return ids


def content_contains_username(content: str | None, username: str | None) -> bool:
    if not content or not username:
        return False
    return username in content


def describe_bot_mention(
    cfg: MyConversationConfig,
    mentioned_user_ids: list[int],
    content: str | None,
) -> tuple[bool, str]:
    if cfg.user_id and any(user_id_matches(cfg.user_id, uid) for uid in mentioned_user_ids):
        return True, "mentioned-user-ids"
    if content_contains_username(content, cfg.username):
        return True, "username-in-content"
    return False, "none"


def should_accept_group_message(
    cfg: MyConversationConfig,
    group_id: int,
    sender_user_id: int,
) -> bool:
    if cfg.user_id and user_id_matches(cfg.user_id, sender_user_id):
        return False
    return should_accept_group(cfg, group_id)


def resolve_mention_gate(
    cfg: MyConversationConfig,
    group_id: int,
    raw_body: str,
    mentioned_user_ids: list[int],
    has_control_command: bool = False,
) -> MentionGateResult:
    if not should_require_mention(cfg, group_id):
        was_mentioned, _ = describe_bot_mention(cfg, mentioned_user_ids, raw_body)
        return MentionGateResult(False, "mention-not-required", was_mentioned)

    was_mentioned, match = describe_bot_mention(cfg, mentioned_user_ids, raw_body)
    if was_mentioned:
        return MentionGateResult(False, "mentioned", True)

    if has_control_command:
        return MentionGateResult(False, "authorized-command", True)

    return MentionGateResult(True, "missing-mention", False)

File: reference/typescript/src/inbound/mention.ts
import {
  type MyConversationChannelConfig,
  mentionedUserIdsInclude,
  shouldAcceptGroup,
  shouldRequireMention,
  userIdMatches,
} from "../config.js";

const MENTION_TOKEN_RE = /\[\[@([^:\]]+):(\d+)\]\]/g;

export type InboundMentionCandidate = {
  groupId: number;
  senderUserId: number;
  content?: string;
  mentionedUserIds?: number[];
};

/** @deprecated Username matching uses config.username directly. */
export type MentionContext = {
  username?: string;
};

export function normalizeMentionedUserIds(mentionedUserIds: number[]): number[] {
  const normalized = new Set<number>();
  for (const id of mentionedUserIds) {
    if (Number.isFinite(id) && id > 0) {
      normalized.add(id);
    }
  }
  return [...normalized];
}

export function stripMentionTokensForCommandDetection(content: string): string {
  return content.replace(MENTION_TOKEN_RE, " ").replace(/\s+/g, " ").trim();
}

export function detectInboundControlCommand(params: {
  rawBody: string;
  cfg: unknown;
  hasControlCommand: (body: string, cfg: unknown) => boolean;
}): boolean {
  const trimmed = params.rawBody.trim();
  if (!trimmed) {
    return false;
  }
  if (params.hasControlCommand(trimmed, params.cfg)) {
    return true;
  }

  const withoutMentions = stripMentionTokensForCommandDetection(trimmed);
  if (withoutMentions && withoutMentions !== trimmed) {
    return params.hasControlCommand(withoutMentions, params.cfg);
  }

  return false;
}

export type MentionGateDebug = {
  requireMention: boolean;
  wasMentioned: boolean;
  mentionMatch: string;
  mentionedUserIds: number[];
  botUserId?: string;
  hasControlCommand: boolean;
  allowTextCommands: boolean;
  commandAuthorized: boolean;
};

export function contentContainsBotUsername(
  content: string | undefined,
  config: MyConversationChannelConfig,
): boolean {
  const username = config.username?.trim();
  if (!username || content == null || content === "") {
    return false;
  }
  return content.includes(username);
}

export function isBotMentionedByUserIds(
  config: MyConversationChannelConfig,
  mentionedUserIds: number[],
  content?: string,
): boolean {
  return describeBotMentionMatch(config, mentionedUserIds, content)
    .wasMentioned;
}

export function describeBotMentionMatch(
  config: MyConversationChannelConfig,
  mentionedUserIds: number[],
  content?: string,
): { wasMentioned: boolean; match: string } {
  if (
    config.userId != null &&
    mentionedUserIdsInclude(config.userId, mentionedUserIds)
  ) {
    return { wasMentioned: true, match: "mentioned-user-ids" };
  }

  if (contentContainsBotUsername(content, config)) {
    return { wasMentioned: true, match: "username-in-content" };
  }

  return { wasMentioned: false, match: "none" };
}

/** Self-message and allowlist checks only — mention gating is handled separately. */
export function shouldAcceptGroupMessage(
  candidate: InboundMentionCandidate,
  config: MyConversationChannelConfig,
): boolean {
  if (
    config.userId != null &&
    userIdMatches(config.userId, candidate.senderUserId)
  ) {
    return false;
  }

  return shouldAcceptGroup(config, candidate.groupId);
}

export function resolveMyConversationMentionGate(params: {
  account: MyConversationChannelConfig;
  groupId: number;
  rawBody: string;
  mentionedUserIds: number[];
  allowTextCommands: boolean;
  hasControlCommand: boolean;
  commandAuthorized: boolean;
}): {
  shouldSkip: boolean;
  reason: string;
  effectiveWasMentioned: boolean;
  debug: MentionGateDebug;
} {
  const mentionedUserIds = normalizeMentionedUserIds(params.mentionedUserIds);
  const requireMention = shouldRequireMention(params.account, params.groupId);
  const mentionMatch = describeBotMentionMatch(
    params.account,
    mentionedUserIds,
    params.rawBody,
  );
  const debug: MentionGateDebug = {
    requireMention,
    wasMentioned: mentionMatch.wasMentioned,
    mentionMatch: mentionMatch.match,
    mentionedUserIds,
    botUserId: params.account.userId,
    hasControlCommand: params.hasControlCommand,
    allowTextCommands: params.allowTextCommands,
    commandAuthorized: params.commandAuthorized,
  };

  if (!requireMention) {
    return {
      shouldSkip: false,
      reason: "mention-not-required",
      effectiveWasMentioned: mentionMatch.wasMentioned,
      debug,
    };
  }

  if (mentionMatch.wasMentioned) {
    return {
      shouldSkip: false,
      reason: "mentioned",
      effectiveWasMentioned: true,
      debug,
    };
  }

  if (
    params.hasControlCommand &&
    params.allowTextCommands &&
    params.commandAuthorized
  ) {
    return {
      shouldSkip: false,
      reason: "authorized-command",
      effectiveWasMentioned: true,
      debug,
    };
  }

  return {
    shouldSkip: true,
    reason: "missing-mention",
    effectiveWasMentioned: false,
    debug,
  };
}

/** @deprecated Use shouldAcceptGroupMessage + resolveMyConversationMentionGate. */
export function shouldHandleInboundMessage(
  candidate: InboundMentionCandidate,
  config: MyConversationChannelConfig,
  _context: MentionContext = {},
): boolean {
  if (!shouldAcceptGroupMessage(candidate, config)) {
    return false;
  }

  if (!shouldRequireMention(config, candidate.groupId)) {
    return true;
  }

  return isBotMentionedByUserIds(
    config,
    normalizeMentionedUserIds(candidate.mentionedUserIds ?? []),
    candidate.content,
  );
}

export function wasInboundBotMentioned(
  rawBody: string | undefined,
  mentionedUserIds: number[] | undefined,
  config: MyConversationChannelConfig,
  _context: MentionContext = {},
): boolean {
  return isBotMentionedByUserIds(
    config,
    normalizeMentionedUserIds(mentionedUserIds ?? []),
    rawBody,
  );
}
